### PR TITLE
refactor: Change catalog configuration so it is entirely dsn based / support end to end testing without postgres

### PR DESCRIPTION
**Pull request by alamb**
# Rationale
Closes https://github.com/influxdata/influxdb_iox/issues/7285

I would like to make it easier to run run end to end  tests (see https://github.com/influxdata/influxdb_iox/issues/7285) and thus permitting the end to end tests to run with sqlite is good

# Changes:
1. Update the catalog configuration to be entirely based on the `INFLUXDB_IOX_CATALOG_DSN` rather than DSN and `INFLUXDB_IOX_CATALOG_TYPE`.
2. Remove `INFLUXDB_IOX_CATALOG_TYPE`

So the DSN can now be one of:
1. postgres (unchanged)
3. `memory` (memory)
4. `sqlite://<filename>`

For end to end testing, it can also be `sqlite` which will result in using a temporary file

Note this does not yet enable the sqlite based catalogs by default in end to end testing (or CI). That is tracked by https://github.com/influxdata/influxdb_iox/issues/7709

# Example uses
Now you can run end to end tests like this (without postgres) and it works great most of the time:

```sql
TEST_INTEGRATION=1 TEST_INFLUXDB_IOX_CATALOG_DSN=sqlite cargo test --test end_to_end 
```

### CODE REVIEW

File: clap_blocks/src/catalog_dsn.rs [hunk 1]
```diff
@@ -50,23 +50,16 @@ fn default_hotswap_poll_interval_timeout() -> &'static str {
...
-    /// Catalog connection string. Required if catalog is set to postgres.
+    /// Catalog connection string.
     ///
-    /// The dsn is interpreted based on the type of catalog used.
+    /// The dsn determines the type of catalog used.
     ///
     /// PostgreSQL: `postgresql://postgres@localhost:5432/postgres`
     ///
-    /// Sqlite (a local filename): '/tmp/foo.sqlite'
+    /// Sqlite (a local filename /tmp/foo.sqlite): `sqlite:///tmp/foo.sqlite`
```
Comment (carols10cents) [inline clap_blocks/src/catalog_dsn.rs:59]: Hm, this says specifying `sqlite` looks like `sqlite:///tmp/foo.sqlite` but the error messages above say it looks like `file:///tmp/catalog.sqlite`? Does the `sqlite` vs `file` protocol matter?

File: clap_blocks/src/catalog_dsn.rs [hunk 2]
```diff
@@ -119,123 +112,59 @@ pub struct CatalogDsnConfig {
...
+            let mem = MemCatalog::new(metrics);
+
+            let mut txn = mem.start_transaction().await.context(CatalogSnafu)?;
+            create_or_get_default_records(txn.deref_mut())
+                .await
+                .context(CatalogSnafu)?;
+            txn.commit().await.context(CatalogSnafu)?;
+
+            Ok(Arc::new(mem) as Arc<dyn Catalog>)
+        } else if let Some(file_path) = dsn.strip_prefix("sqlite://") {
```
Comment (carols10cents) [inline clap_blocks/src/catalog_dsn.rs:154]: Aha, looks like `sqlite://` is required, so the error message text needs to be changed.

File: test_helpers_end_to_end/src/config.rs [hunk 3]
```diff
@@ -310,3 +325,24 @@ fn random_catalog_schema_name() -> String {
         .map(char::from)
         .collect::<String>()
 }
+
+/// Rewrites the special "iox_test_sqlite" catalog DSN to a new
```
Comment (carols10cents) [inline test_helpers_end_to_end/src/config.rs:329]: What is the special "iox_test_sqlite" catalog DSN? I don't see that anywhere else in this PR or in main. 

On line 337 below, it's checking for the exact string `"sqlite"`, is that what this comment should say? 

Are there any tests that currently specify `"sqlite"` as the dsn in the `TestConfig`?

**Review by carols10cents — CHANGES_REQUESTED** [review]
I think some comments and error messages that need to be updated to be consistent with the code and each other? 

This looks like an awesome simplification!

File: test_helpers_end_to_end/src/config.rs (hunk 3 above)
Comment (alamb) [reply test_helpers_end_to_end/src/config.rs:329]: this is a good catch -- it is a left over from a prior implementation -- I will update the docs / comments / error messges

File: clap_blocks/src/catalog_dsn.rs (hunk 2 above)
Comment (alamb) [reply clap_blocks/src/catalog_dsn.rs:154]: Fixed in 70385eeb0

File: test_helpers_end_to_end/src/config.rs (hunk 3 above)
Comment (alamb) [reply test_helpers_end_to_end/src/config.rs]: 70385eeb0

**Review by carols10cents — APPROVED** [review]
Looks great now, thank you!

**Review by domodwyer — APPROVED** [review]

File: clap_blocks/src/catalog_dsn.rs [hunk 4]
```diff
@@ -50,23 +50,16 @@ fn default_hotswap_poll_interval_timeout() -> &'static str {
 /// CLI config for catalog DSN.
 #[derive(Debug, Clone, Default, clap::Parser)]
 pub struct CatalogDsnConfig {
-    /// The type of catalog to use. "memory" is only useful for testing purposes.
```
[the remark this answers is not in the document]
Comment (alamb) [reply clap_blocks/src/catalog_dsn.rs:53]: I also checked a local checkout of `k8-idpe`:

```
-*- mode: grep; default-directory: "~/Software/k8s-idpe/" -*-
Grep started at Tue May  2 06:38:36

rg -n -H --no-heading -e 'INFLUXDB_IOX_CATALOG_TYPE' $(git rev-parse --show-toplevel || pwd)
```

So I am going to claim this is good to go

**Comment by domodwyer** [comment]
I have reverted this change ⚠️

**Pull request merged**